src/vm_pages.rs

```diff
@@ -1131,38 +1132,57 @@ impl<'a, T: GuestStagePagingMode> FinalizedVmPages<'a, T> {
         if !state_pages.is_contiguous() {
             return Err(Error::NonContiguousPages);
         }
-        let vcpu_pages = self.get_converted_pages(vcpus_addr, num_vcpu_pages)?;
-        if !vcpu_pages.is_contiguous() {
-            return Err(Error::NonContiguousPages);
-        }
         let id = self
             .inner
             .page_tracker
             .add_active_guest()
             .map_err(Error::GuestId)?;
-
         let guest_root_pages =
             SequentialPages::from_pages(self.assign_state_pages_for(guest_root_pages, id)).unwrap();
         let guest_root =
             GuestStagePageTable::new(guest_root_pages, id, self.inner.page_tracker.clone())
                 .unwrap();
 
-        let mut state_pages = self.assign_state_pages_for(state_pages, id);
-        let box_page = state_pages.next().unwrap();
-        let region_vec_pages = SequentialPages::from_pages(state_pages).unwrap();
+        let mut state_pages =
+            SequentialPages::from_pages(self.assign_state_pages_for(state_pages, id)).unwrap();
+        let vm_state_pages = state_pages
+            .take(TVM_VM_STATE_PAGES)
+            .ok_or(Error::NotEnoughMemory)?;
+        let region_vec_pages = state_pages;
         let region_vec = VmRegionList::new(region_vec_pages, self.inner.page_tracker.clone());
 
+        let vm = Vm::new(
+            VmPages::new(guest_root, region_vec, self.inner.nesting + 1),
+            VmCpus::new(),
+        )
+        .map_err(Error::VmCreationFailed)?;
+
+        Ok((vm, vm_state_pages))
+    }
+
+    pub fn alloc_vcpu(
+        &self,
+        vcpu_id: u64,
+        shared_area: VmCpuSharedArea,
+        state_addr: GuestPageAddr,
+        to: InitializingVmPages<T>,
+    ) -> Result<PageBox<Mutex<VmCpu>>> {
+        let converted_pages = self.get_converted_pages(state_addr, VM_CPU_STATE_PAGES)?;
+        if !converted_pages.is_contiguous() {
+            return Err(Error::NonContiguousPages);
+        }
+
+        let new_owner = to.page_owner_id();
         let vcpu_pages =
-            SequentialPages::from_pages(self.assign_state_pages_for(vcpu_pages, id)).unwrap();
+            SequentialPages::from_pages(self.assign_state_pages_for(converted_pages, new_owner))
+                .unwrap();
```

Comment:
Is this unwrap guaranteed never to panic? If so, why? Can it instead return an error?

Reply:
I was going with the flow of the rest of Salus code.

Not a complete expert on the project yet, but since the pages comes from converted_pages _should_ not.

I think this has the same possibility of being wrong as the rest of the salus code.

The question of whether carefully selected page list passed by the TEE host might or might not crash salus -- we should make sure that it doesn't -- is definitely something that caught my eye.
I'd like to volunteer for that, but I think this is beyond the scope of this change.

Reply:
It is guaranteed to succeed because of the `is_contiguous()` check above. The reason the check is separate from the creation of the `SequentialPages` is to avoid having to unwind the work done by `assign_state_pages_for()`. This deserves a comment, which I should've added when I initially introduced this pattern in `create_guest_vm()` (whoops)